src/dos/program_more.cpp

```diff
@@ -0,0 +1,508 @@
+{
+	// We need to be able to read STDIN for key presses, but it is most
+	// likely redirected - so clone the handle, and reconstruct real STDIN
+	// from STDERR (idea from FreeDOS implementation,
+	// https://github.com/FDOS/more/blob/master/src/more.c)
+	if (!DOS_DuplicateEntry(STDIN, &input_handle) ||
+	    !DOS_ForceDuplicateEntry(STDERR, STDIN)) {
+		LOG_ERR("DOS: Unable to prepare handles in MORE.COM");
+		return;
+	}
+
+	WriteOut("\n");
+	WriteOut(MSG_Get("SHELL_CMD_MORE_START"));
+	WriteOut("\n");
+
+	// Since this CAN be STDIN input (there is no way to check),
+	// enable emergency break sequence while processing
+	break_enable = true;
+	DisplaySingleStream();
+}
+
+MORE::Decision MORE::DisplaySingleStream()
+{
+	auto previous_column = GetCurrentColumn();
+
+	break_idx    = 0;
+	tab_left     = 0;
+	skip_next_cr = false;
+	skip_next_lf = false;
+
+	auto decision = Decision::NextFile;
+	while (true) {
+		if (shutdown_requested) {
+			decision = Decision::Terminate;
+			break;
+		}
+
+		// Read character
+		char code = 0;
+		if (!GetCharacter(code)) {
+			decision = Decision::NextFile; // end of file
+			break;
+		}
+
+		// A trick to make it more resistant to ANSI cursor movements
+		const auto current_row = GetCurrentRow();
+		if (line_counter > current_row)
+			line_counter = current_row;
+
+		// Handle new line characters
+		bool new_line = false;
+		if (code == code_cr) {
+			skip_next_lf = true;
+			new_line     = true;
+		} else if (code == code_lf) {
+			skip_next_cr = true;
+			new_line     = true;
+		} else {
+			skip_next_cr = false;
+			skip_next_lf = false;
+		}
+
+		// Duplicate character on the output
+		if (new_line)
+			code = '\n';
+		WriteOut("%c", code);
+
+		// Detect 'new line' due to character passing the last column
+		const auto current_column = GetCurrentColumn();
+		if (!current_column && previous_column)
+			new_line = true;
+		previous_column = current_column;
+
+		// Update new line counter, decide if pause needed
+		if (new_line && current_row)
+			++line_counter;
+		if (line_counter < max_lines)
+			continue;
+
+		// New line occured just enough times for a pause
+		decision = PromptUser();
+		if (decision == Decision::Terminate || decision == Decision::NextFile)
+			break;
+	}
+
+	if (GetCurrentColumn()) {
+		++line_counter;
+		WriteOut("\n");
+	}
+
+	return decision;
+}
+
+MORE::Decision MORE::PromptUser()
+{
+	line_counter = 0;
+	const bool multiple_files = input_files.size() > 1;
+
+	if (GetCurrentColumn())
+		WriteOut("\n");
+
+	if (multiple_files)
+		WriteOut(MSG_Get("SHELL_CMD_MORE_PROMPT_MULTI"));
+	else
+		WriteOut(MSG_Get("SHELL_CMD_MORE_PROMPT_SINGLE"));
+
+	auto decision = Decision::Terminate;
+	while (!shutdown_requested) {
+		CALLBACK_Idle();
+
+		uint16_t count = 1;
+		char choice    = 0;
+		DOS_ReadFile(STDIN, reinterpret_cast<uint8_t *>(&choice), &count);
+
+		if (count == 0 || choice == code_ctrl_c || choice == code_esc ||
+		    choice == 'q' || choice == 'Q')
+			break;
```

Comment:
These choices feel like the user is asking for a `Decision::Terminate` (so I wondered why it wasn't being set here.. so I had to figure out that it's the the default further up.  Is there any way to set the Terminate decision here?  This would logically follow the same approach for the 'NextFile' decision)

Reply:
Added setting the variable. I hope this won't cause compiler warnings.